exoscale_auth.py

```diff
@@ -29,3 +31,55 @@ def __call__(self, request):
             'Authorization': auth,
         })
         return request
+
+class ExoscaleV2Auth(AuthBase):
+    def __init__(self, key, secret):
+        self.key = key
+        self.secret = secret.encode('utf-8')
+
+    def __call__(self, request):
+        auth_header = u'EXO2-HMAC-SHA256 credential={}'.format(self.key)
+
+        # Request method/URL path
+        msg = '{method} {path}\n'.format(
+            method=request.method,
+            path=urlparse(request.url).path)
+
+        # Request body
+        if request.body:
+            msg += body.encode('utf-8')
+        msg += '\n'
+
+        # Request query string parameters
+        # Important: this is order-sensitive, we have to have to sort
+        # parameters alphabetically to ensure signed # values match the
+        # names listed in the "signed-query-args=" signature pragma.
+        params = parse_qs(urlparse(request.url).query)
+        signed_params = sorted(params.keys())
+        for p in signed_params:
+            if len(params[p]) != 1:
+                continue
+            msg += params[p][0]
+        msg += '\n'
+        if len(signed_params) > 0:
+            auth_header += ',signed-query-args={}'.format(';'.join(signed_params))
+
+        # Request headers -- none at the moment
+        # Note: the same order-sensitive caution for query string parameters
+        # applies to headers.
+        msg += '\n'
```

Comment:
Instead of managing the `\n`s, you could prepare a list of strings and join at the end, e.g.:

```python
parts = [method_and_path, body, params, headers]
msg = "\n".join(parts)
```

This involves less string mutating and is perhaps a bit easier to read (it can be subjective :smile_cat:)

Reply:
Good idea, changed in d05e994.